### PR TITLE
IGNITE-23516 : Calcite. Unify rounding of numerics.

### PR DESCRIPTION
**Pull request by Vladsz83**

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java [hunk 1]
```diff
@@ -308,17 +297,20 @@ public static short convertToShortExact(long x) {
 
     /** Cast value to {@code short}, throwing an exception if the result overflows an {@code short}. */
     public static short convertToShortExact(double x) {
-        if (x > Short.MAX_VALUE || x < Short.MIN_VALUE)
```
Comment (alex-plekhanov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java:311]:
```
x = round(x);
```
And left the old code?
In this case we don't need arrays SHORT_BOUNDS and other.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java [hunk 2]
```diff
@@ -308,17 +297,20 @@ public static short convertToShortExact(long x) {
...
+        if (x < (Double)SHORT_BOUNDS[2] || x > (Double)SHORT_BOUNDS[3])
             throw new ArithmeticException(SMALLINT.getName() + " overflow");
 
-        return (short)x;
+        return (short)round(x);
     }
 
     /** Cast value to {@code short}, throwing an exception if the result overflows an {@code short}. */
     public static short convertToShortExact(Number x) {
-        checkNumberLongBounds(SMALLINT, x);
```
Comment (alex-plekhanov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java:319]: Perhaps we can round x, check bounds for long value, convert it to long and leave convertToShortExact? In this case we don't need checkShortBounds and other boilerplate methods.
Something like this:
```
x = round(x);
checkNumberLongBounds(SMALLINT, x);
return convertToShortExact(x.longValue());
```
Where `round`:
```
    private static Number round(Number x) {
        return convertToBigDecimal(x).setScale(0, NUMERIC_ROUNDING_MODE);
    }
```

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java [hunk 3]
```diff
@@ -333,16 +325,142 @@ public static byte convertToByteExact(long x) {
...
+
+            if (longVal >= Byte.MIN_VALUE && longVal <= Byte.MAX_VALUE)
+                return null;
+        }
+
+        throw new ArithmeticException(TINYINT.getName() + " overflow");
+    }
+
+    /** */
+    private static double round(double x) {
```
Comment (alex-plekhanov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java:463]: Perhaps `round` is not a good name. Becouse it prepares to round, but not do round itself.

Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java:463]: Didn't get it. it does round. Changes the value just like HALF_UP strategy.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java (hunk 1 above)
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java:311]: The overflow exception might be thrown when must not be. 32767.1 is ok. Will be casted to 32767 with the new code. But the overflow-exception is thrown with the old code.

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/DataTypesTest.java [hunk 4]
```diff
@@ -43,6 +44,74 @@
...
+        for (String numTypeName : F.asList("DOUBLE", "FLOAT")) {
+            lst.add(F.asList(numTypeName, floatingVal(1.4999f, numTypeName), "DECIMAL(1)", new BigDecimal(1)));
+            lst.add(F.asList(numTypeName, floatingVal(-1.4999f, numTypeName), "DECIMAL(1)", new BigDecimal(-1)));
+            lst.add(F.asList(numTypeName, floatingVal(1.5f, numTypeName), "DECIMAL(1)", new BigDecimal(2)));
+            lst.add(F.asList(numTypeName, floatingVal(-1.5f, numTypeName), "DECIMAL(1)", new BigDecimal(-2)));
+
+            lst.add(F.asList(numTypeName, floatingVal(1.4999f, numTypeName), "BIGINT", 1L));
+            lst.add(F.asList(numTypeName, floatingVal(-1.4999f, numTypeName), "BIGINT", -1L));
+            lst.add(F.asList(numTypeName, floatingVal(1.5f, numTypeName), "BIGINT", 2L));
+            lst.add(F.asList(numTypeName, floatingVal(-1.5f, numTypeName), "BIGINT", -2L));
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/DataTypesTest.java:89]: see `testCastDecimalOverflows()`

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/DataTypesTest.java [hunk 5]
```diff
@@ -43,6 +44,74 @@
...
+
+            lst.add(F.asList(numTypeName, floatingVal(1.4999f, numTypeName), "SMALLINT", (short)1));
+            lst.add(F.asList(numTypeName, floatingVal(-1.4999f, numTypeName), "SMALLINT", (short)-1));
+            lst.add(F.asList(numTypeName, floatingVal(1.5f, numTypeName), "SMALLINT", (short)2));
+            lst.add(F.asList(numTypeName, floatingVal(-1.5f, numTypeName), "SMALLINT", (short)-2));
+
+            lst.add(F.asList(numTypeName, floatingVal(1.4999f, numTypeName), "TINYINT", (byte)1));
+            lst.add(F.asList(numTypeName, floatingVal(-1.4999f, numTypeName), "TINYINT", (byte)-1));
+            lst.add(F.asList(numTypeName, floatingVal(1.5f, numTypeName), "TINYINT", (byte)2));
+            lst.add(F.asList(numTypeName, floatingVal(-1.5f, numTypeName), "TINYINT", (byte)-2));
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/DataTypesTest.java:104]: I'll check but this case is covered in `testCastDecimalOverflows()`

[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/DataTypesTest.java:104]: It does throw `Caused by: java.lang.ArithmeticException: TINYINT overflow`. Are you sure you look ath the test with 'sqlTxMode=NONE'?

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java (hunk 2 above)
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java:319]: Attached v2 : https://github.com/apache/ignite/pull/11644

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java (hunk 3 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java:463]: Method doesn't round value. It's adds or substracts 0.5 to the value (prepare to round). round method should return value without floating point.

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/DataTypesTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/DataTypesTest.java:106]: Yes, with sqlTxMode=NONE, testRoundingOfNumericsPrecasted, testRoundingOfDynamicNumerics, testRoundingOfDynamicNumericsPrecasted passed, only testRoundingOfNumerics throws correct exception.

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/DataTypesTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/DataTypesTest.java:91]: testCastDecimalOverflows only check literals, but not dynamic parameters.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java (hunk 1 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java:311]: New code still has errors on bounds.
Condition can be changed to: `if (x >= Short.MAX_VALUE + 1d || x =< Short.MIN_VALUE - 1d)`

Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java:311]: New code still has errors on bounds.
Condition can be changed to: `if (x >= Short.MAX_VALUE + 1d || x =< Short.MIN_VALUE - 1d)`

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java (hunk 3 above)
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java]: Renamed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java (hunk 1 above)
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java:311]: Was fixed. Lost in commits. Pushed.

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/DataTypesTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/DataTypesTest.java:106]: Fixed

**Comment by Vladsz83** [comment]
#implemented in 11644

**Pull request closed without merging**